enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/TriggersAnnotationEnricher.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2016 Red Hat, Inc.
+ *
+ * Red Hat licenses this file to you under the Apache License, version
+ * 2.0 (the "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package io.fabric8.maven.enricher.standard;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import io.fabric8.kubernetes.api.builder.TypedVisitor;
+import io.fabric8.kubernetes.api.builder.Visitable;
+import io.fabric8.kubernetes.api.model.Container;
+import io.fabric8.kubernetes.api.model.ContainerBuilder;
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.KubernetesListBuilder;
+import io.fabric8.kubernetes.api.model.extensions.DaemonSetBuilder;
+import io.fabric8.kubernetes.api.model.extensions.ReplicaSetBuilder;
+import io.fabric8.kubernetes.api.model.extensions.StatefulSetBuilder;
+import io.fabric8.maven.core.util.Configs;
+import io.fabric8.maven.core.util.JSONUtil;
+import io.fabric8.maven.docker.util.ImageName;
+import io.fabric8.maven.enricher.api.BaseEnricher;
+import io.fabric8.maven.enricher.api.EnricherContext;
+import io.fabric8.openshift.api.model.ImageChangeTrigger;
+import io.fabric8.openshift.api.model.ImageChangeTriggerBuilder;
+
+import java.util.*;
+
+/**
+ * This adds a `image.openshift.io/triggers` tag to all kubernetes resources in order to make them run on Openshift when using ImageStreams.
+ *
+ * @author nicola
+ * @since 10/05/18
+ */
+public class TriggersAnnotationEnricher extends BaseEnricher {
+
+    private static final String TRIGGERS_ANNOTATION = "image.openshift.io/triggers";
+
+    // Available configuration keys
+    private enum Config implements Configs.Key {
+
+        /**
+         * Comma-separated list of container names that should be enriched (default all that apply)
+         */
+        containers;
+
+        protected String d; public String def() {
+            return d;
+        }
+    }
+
+
+    public TriggersAnnotationEnricher(EnricherContext buildContext) {
+        super(buildContext, "fmp-triggers-annotation");
+    }
+
+    @Override
+    public void adapt(KubernetesListBuilder builder) {
+
+        builder.accept(new TypedVisitor<StatefulSetBuilder>() {
+            @Override
+            public void visit(StatefulSetBuilder o) {
+                if (canWriteTriggers(o.build())) {
```

Comment:
refactor: instead of repeating add trigger logic on three places, can we extract it in a generic method?

like this or maybe using another way

```java
public void visit(StatefulSetBuilder statefulSetBuilder) {
  addImageTrigger(statefulSetBuilder);
}

void <V exteands VisitableBuilder> addImageTrigger(V  builder) {
.......
}
```

Reply:
I tried to do it, but the `editOrNewMetadata()` and other methods operating on metadata are specific of each builder and do not have a super-class/interface, so even if the syntax is exactly the same, there's no interface containing them (and also return types are different in all three cases).

Reply:
Still I can refactor it a bit..